src/util/recentlyEditedTree.js

```diff
@@ -0,0 +1,153 @@
+import _ from 'lodash'
+import { timestamp } from './timestamp'
+import { pathToContext } from './pathToContext'
+import { timeDifference } from '../util.js'
+import { EM_TOKEN } from '../constants'
+import { produce } from 'immer'
+
+const EDIT_TIME_MAX = 1000 // time diff limit in second for replacing descendants by ancestor
+
+const findTreeDeepestSubcontext = (tree, path, index = 0) => {
+  const node = tree[path[index]]
+  if (node) return findTreeDeepestSubcontext(node, path, index + 1)
+  else return { node: index === 0 ? null : tree, path: path.slice(0, index) }
+}
+
+export const findTreeDescendants = (tree, startingPath) => {
+  const node = _.get(tree, startingPath)
+  if (!node) return []
+  else if (node.leaf) return [{ ...node }]
+  else return Object.keys(node).reduce((acc, child) => [...acc, ...findTreeDescendants(tree, [...startingPath, child])], [])
+}
```

Comment:
This still needs some work. Notice that you are still calling `_.get` each iteration. This results in complexity of `O(depth) * O(depth)`. `_.get` is great when a single node is accessed, but it shouldn't be used in a loop or iteration, since it will repeat work. Instead of you want a `O(n)` recursive traversal function. Let me know if you need more guidance on this.

Reply:
Yes understood. Done!